firebase-firestore/src/main/java/com/google/firebase/firestore/remote/WatchChangeAggregatorTestingHooks.java

```diff
@@ -0,0 +1,110 @@
+import com.google.firebase.firestore.util.Executors;
+import java.util.HashMap;
+import java.util.Map;
+
+final class WatchChangeAggregatorTestingHooks {
+
+  private WatchChangeAggregatorTestingHooks() {}
+
+  private static final Map<Object, ExistenceFilterMismatchListener>
+      existenceFilterMismatchListeners = new HashMap<>();
+
+  /**
+   * Notifies all registered {@link ExistenceFilterMismatchListener}` listeners registered via
+   * {@link #addExistenceFilterMismatchListener}.
+   *
+   * @param info Information about the existence filter mismatch to deliver to the listeners.
+   */
+  static void notifyOnExistenceFilterMismatch(ExistenceFilterMismatchInfo info) {
+    synchronized (existenceFilterMismatchListeners) {
+      for (ExistenceFilterMismatchListener listener : existenceFilterMismatchListeners.values()) {
+        Executors.BACKGROUND_EXECUTOR.execute(() -> listener.onExistenceFilterMismatch(info));
+      }
+    }
+  }
+
+  /**
+   * Registers a {@link ExistenceFilterMismatchListener} to be notified when an existence filter
+   * mismatch occurs in the Watch listen stream.
+   *
+   * <p>The relative order in which callbacks are notified is unspecified; do not rely on any
+   * particular ordering. If a given callback is registered multiple times then it will be notified
+   * multiple times, once per registration.
+   *
+   * <p>The thread on which the callback occurs is unspecified; listeners should perform their work
+   * as quickly as possible and return to avoid blocking any critical work. In particular, the
+   * listener callbacks should <em>not</em> block or perform long-running operations. Listener
+   * callbacks can occur concurrently with other callbacks on the same and other listeners.
+   *
+   * @param listener the listener to register.
+   * @return an object that unregisters the given listener via its {@link
+   *     ListenerRegistration#remove} method; only the first unregistration request does anything;
+   *     all subsequent requests do nothing.
+   */
+  @VisibleForTesting
+  static ListenerRegistration addExistenceFilterMismatchListener(
+      @NonNull ExistenceFilterMismatchListener listener) {
+    checkNotNull(listener, "a null listener is not allowed");
+
+    Object listenerId = new Object();
+    synchronized (existenceFilterMismatchListeners) {
+      existenceFilterMismatchListeners.put(listenerId, listener);
+    }
+
+    return () -> {
+      synchronized (existenceFilterMismatchListeners) {
+        existenceFilterMismatchListeners.remove(listenerId);
+      }
+    };
+  }
+
+  interface ExistenceFilterMismatchListener {
+    @AnyThread
+    void onExistenceFilterMismatch(ExistenceFilterMismatchInfo info);
+  }
+
+  @AutoValue
+  abstract static class ExistenceFilterMismatchInfo {
+
+    static ExistenceFilterMismatchInfo create(int localCacheCount, int existenceFilterCount) {
+      return new AutoValue_WatchChangeAggregatorTestingHooks_ExistenceFilterMismatchInfo(
```

Comment:
is `AutoValue_WatchChangeAggregatorTestingHooks_ExistenceFilterMismatchInfo` a class? Is it generated by the AutoValue library? why we wan to use it here???

Reply:
Yes, `AutoValue_WatchChangeAggregatorTestingHooks_ExistenceFilterMismatchInfo` is indeed a class that is generated by the "AutoValue" library: https://github.com/google/auto/blob/main/value/userguide/index.md

AutoValue is an idomatic way to create "data classes" in Java at Google. It saves writing the boilerplate of setters, getters, builders, etc. It does indeed look like overkill here since the class only contains two integers; however, in the mila/BloomFilter branch it makes a little more sense because it also contains the bloom filter parameters: https://github.com/firebase/firebase-android-sdk/blob/971632689bc363743b75c6cf6baf25bdd0317397/firebase-firestore/src/main/java/com/google/firebase/firestore/remote/WatchChangeAggregatorTestingHooks.java#L128-L164

We also use AutoValue in a few places in the Firestore Android SDK, such as here: https://github.com/firebase/firebase-android-sdk/blob/687e079d401db1a75d2559038b879726a476ca3e/firebase-firestore/src/main/java/com/google/firebase/firestore/model/FieldIndex.java#L36-L37 and https://github.com/firebase/firebase-android-sdk/blob/687e079d401db1a75d2559038b879726a476ca3e/firebase-firestore/src/main/java/com/google/firebase/firestore/model/FieldIndex.java#L68-L69